src/status.rs

```diff
@@ -1,166 +1,156 @@
-use std::fmt;
-
+use bytes::Bytes;
 use h2;
-use http::header::HeaderValue;
+use http::HeaderMap;
+use http::header::{HeaderValue, InvalidHeaderValueBytes};
 
 #[derive(Debug, Clone)]
 pub struct Status {
     code: Code,
+    error_message: Option<Bytes>,
+    binary_error_details: Option<Bytes>,
 }
 
-#[derive(Clone, Copy, PartialEq, Eq)]
-pub struct Code(Code_);
-
 impl Status {
-    #[inline]
-    pub fn code(&self) -> Code {
-        self.code
+    pub fn with_code(code: Code) -> Status {
+        Status {
+            code,
+            error_message: None,
+            binary_error_details: None,
+        }
     }
 
-    pub const OK: Status = Status {
-        code: Code(Code_::Ok),
-    };
-
-    pub const CANCELED: Status = Status {
-        code: Code(Code_::Canceled),
-    };
-
-    pub const UNKNOWN: Status = Status {
-        code: Code(Code_::Unknown),
-    };
-
-    pub const INVALID_ARGUMENT: Status = Status {
-        code: Code(Code_::InvalidArgument),
-    };
-
-    pub const DEADLINE_EXCEEDED: Status = Status {
-        code: Code(Code_::DeadlineExceeded),
-    };
-
-    pub const NOT_FOUND: Status = Status {
-        code: Code(Code_::NotFound),
-    };
-
-    pub const ALREADY_EXISTS: Status = Status {
-        code: Code(Code_::AlreadyExists),
-    };
-
-    pub const PERMISSION_DENIED: Status = Status {
-        code: Code(Code_::PermissionDenied),
-    };
-
-    pub const RESOURCE_EXHAUSTED: Status = Status {
-        code: Code(Code_::ResourceExhausted),
-    };
-
-    pub const FAILED_PRECONDITION: Status = Status {
-        code: Code(Code_::FailedPrecondition),
-    };
+    pub fn from_header_map(header_map: &mut HeaderMap) -> Option<Status> {
```

Comment:
I'm not sure how I feel about this function taking the header map mutably, and removing the headers from it...that seems somewhat surprising. If I were to use that `HeaderMap` again later, I'd expect all of its headers to still be intact. Perhaps this should take the header map by an immutable reference, and clone the appropriate header values?

Admittedly, removing them from the map may be more performant, but it seems like it could lead to bugs...

Reply:
Yeah, I was pretty torn on this... Switched to immutable reference + clone